internal/ias/client.go

```diff
@@ -27,74 +29,172 @@ func NewIasClient(iasTenantUrl, user, password string) (Client, error) {
 		tenantUrl: iasTenantUrl,
 	}, nil
 }
 
 type client struct {
-	api       *api.ClientWithResponses
+	api       api.ClientWithResponsesInterface
 	tenantUrl string
 }
 
-// TODO: Add logic to handle failed reconciliation correctly without creating new application/secret
-// TODO: Add unit tests with mocked IAS API
-// CreateApplication creates an application in IAS
+// CreateApplication creates an application in IAS. This function is not idempotent, because if an application with the specified
+// name already exists, it will be deleted and recreated.
 func (c client) CreateApplication(ctx context.Context, name string) (Application, error) {
 
-	newApplication := newIasApplication(name)
-	createAppResponse, err := c.api.CreateApplicationWithResponse(ctx, &api.CreateApplicationParams{}, newApplication)
+	existingApp, err := c.getApplicationByName(ctx, name)
 	if err != nil {
 		return Application{}, err
 	}
 
-	// TODO: Do we need handling for 429 (Too Many Requests) or other status codes?
-	if createAppResponse.StatusCode() != http.StatusCreated {
-		ctrl.Log.Error(err, "Failed to create application", "name", name, "statusCode", createAppResponse.StatusCode())
-		return Application{}, fmt.Errorf("failed to create application")
+	// To simplify the logic, if an application with this name already exists, we always delete the application and create
+	// a new one, otherwise we would have to check where the application creation failed and continue at this point.
+	if existingApp != nil {
+		res, err := c.api.DeleteApplicationWithResponse(ctx, *existingApp.Id)
+		if err != nil {
+			return Application{}, err
+		}
+		if res.StatusCode() != http.StatusOK {
+			ctrl.Log.Error(err, "Failed to delete existing application", "id", *existingApp.Id, "statusCode", res.StatusCode())
+			return Application{}, errors.New("failed to delete existing application before creation")
+		}
 	}
 
-	appId, err := extractApplicationId(createAppResponse)
+	appId, err := c.createNewApplication(ctx, name)
 	if err != nil {
 		return Application{}, err
 	}
 	ctrl.Log.Info("Created application", "name", name, "id", appId)
 
-	createSecretResponse, err := c.api.CreateApiSecretWithResponse(ctx, appId, newSecretRequest())
+	clientSecret, err := c.createSecret(ctx, appId)
+	if err != nil {
+		return Application{}, err
+	}
+
+	clientId, err := c.getClientId(ctx, appId)
 	if err != nil {
 		return Application{}, err
 	}
 
-	if createSecretResponse.StatusCode() != http.StatusCreated {
-		ctrl.Log.Error(err, "Failed to create api secret", "id", appId, "statusCode", createSecretResponse.StatusCode())
-		return Application{}, fmt.Errorf("failed to create api secret")
+	return NewApplication(clientId, clientSecret, c.tenantUrl), nil
+}
+
+// DeleteApplication deletes an application in IAS. If the application does not exist, this function does nothing.
+func (c client) DeleteApplication(ctx context.Context, name string) error {
+	existingApp, err := c.getApplicationByName(ctx, name)
```

Comment:
it'd be better to ignore the error if the application is not found. I assume we get an error unsell an app found.